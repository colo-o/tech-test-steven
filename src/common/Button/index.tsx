import { forwardRef, ReactNode, MouseEvent } from 'react';
import classNames from 'classnames';

type ButtonProps = {
  children?: ReactNode;
  onClick?: (ev: MouseEvent<HTMLButtonElement>) => void;
  className?: string;
  disabled?: boolean;
};

type ButtonRef = HTMLButtonElement | null;

const Button = forwardRef<ButtonRef, ButtonProps>(
  ({ children, className, onClick, disabled }, ref) => {
    const buttonClassName = classNames(
      `${className} bg-primary border-2 border-primary rounded-lg px-4 py-2 text-white font-medium box-border cursor-pointer shadow-lg shadow-primary/30 ease-out duration-700`,
      {
        'disabled:bg-transparent disabled:border-gray disabled:text-gray disabled:cursor-not-allowed disabled:shadow-none':
          disabled,
      }
    );

    return (
      <button
        className={buttonClassName}
        ref={ref}
        onClick={onClick}
        disabled={disabled}
      >
        {children}
      </button>
    );
  }
);

Button.displayName = 'Main Button';

export default Button;
