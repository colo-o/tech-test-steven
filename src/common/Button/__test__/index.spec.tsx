import { render, screen, fireEvent } from '@testing-library/react';

import Button from 'common/Button';

describe('[BUTTON]', () => {
  it('Should trigger onClick action', () => {
    const onClickMock = jest.fn();

    render(<Button onClick={onClickMock}>Hello world</Button>);

    const button = screen.getByRole('button', { name: 'Hello world' });
    fireEvent.click(button);

    expect(onClickMock).toBeCalledTimes(1);
    expect(button).toMatchSnapshot();
  });

  it('Should display a disabled button', () => {
    const onClickMock = jest.fn();

    render(
      <Button onClick={onClickMock} disabled>
        Hello world
      </Button>
    );

    const button = screen.getByRole('button', { name: 'Hello world' });
    fireEvent.click(button);

    expect(onClickMock).toBeCalledTimes(0);
    expect(button).toMatchSnapshot();
  });
});
