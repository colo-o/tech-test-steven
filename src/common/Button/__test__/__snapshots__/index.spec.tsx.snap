// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[BUTTON] Should display a disabled button 1`] = `
<button
  class="undefined bg-primary border-2 border-primary rounded-lg px-4 py-2 text-base md:text-lg text-white font-medium box-border cursor-pointer shadow-lg shadow-primary/30 ease-out duration-700 disabled:bg-transparent disabled:border-gray disabled:text-gray disabled:cursor-not-allowed disabled:shadow-none"
  disabled=""
>
  Hello world
</button>
`;

exports[`[BUTTON] Should trigger onClick action 1`] = `
<button
  class="undefined bg-primary border-2 border-primary rounded-lg px-4 py-2 text-base md:text-lg text-white font-medium box-border cursor-pointer shadow-lg shadow-primary/30 ease-out duration-700"
>
  Hello world
</button>
`;
